components/select/Select.razor.cs

```diff
@@ -84,39 +119,52 @@ public IEnumerable<TItem> DataSource
                     SelectedOptionItems.Clear();
 
                     Value = default;
-                    var sameObject = object.ReferenceEquals(_datasource, value);
 
                     _datasource = value;
+                    _dataSourceShallowCopy = value.ToList();
 
-                    if (!sameObject)
-                        OnDataSourceChanged?.Invoke();
+                    OnDataSourceChanged?.Invoke();
 
                     return;
                 }
 
                 if (value != null)
                 {
-                    bool hasChanged;
-
                     if (_datasource == null)
                     {
-                        hasChanged = true;
+                        _dataSourceHasChanged = true;
                     }
                     else
                     {
-                        hasChanged = !value.SequenceEqual(_datasource);
+                        _dataSourceHasChanged = !value.SequenceEqual(_dataSourceShallowCopy, DataSourceEqualityComparer);
                     }
 
-                    if (hasChanged)
+                    if (_dataSourceHasChanged)
                     {
                         OnDataSourceChanged?.Invoke();
-
                         _datasource = value;
+                        if (IsTItemPrimitive)
+                        {
+                            _dataSourceShallowCopy = _datasource.ToList();
+                        }
+                        else
+                        {
+                            var cloneMethod = GetDataSourceItemCloneMethod();
+                            _dataSourceShallowCopy = _datasource.Select(x => (TItem)cloneMethod.Invoke(x, null)).ToList();
```

Comment:
When use default DataSourceEqualityComparer and TItem is not primitive, _dataSourceHasChanged is always true and OnDataSourceChanged will be triggered.

Reply:
Please check commit af8d541 - I added internal `DateSourceEqualityComparer` that is comparing only values of label & value properties of the `DataSource`. Also, I included a test to cover the scenario.